### ISSUE TITLE
[no-explicit-any] Option to disable rule for generic constraints

### ISSUE THREAD

**Issue by anilanar** [typescript-eslint/typescript-eslint#642]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-explicit-any": ["error", { "ignoreGenericConstraints": true }]
  }
}
```

```TS
type IsString<T extends any> = T extends string ? true : false;

type CodomainOf<T extends Record<string, any>> = T extends Record<string, infer R> ? R : never;
```

**Expected Result**
No errors for explicit `any` if they are used in generic constraints.

**Actual Result**
There is no such option to disable errors/warnings for explicit `any` usage in generic constraints.

**Comment by j-f1** [comment]
Does `unknown` have the same behavior here?

**Comment by bradzacher** [comment]
```diff
- type IsString<T extends any> = T extends string ? true : false;
+ type IsString<T> = T extends string ? true : false;
```

Nitpicking, but `extends any` is a useless qualifier.

```diff
- type CodomainOf<T extends Record<string, any>> = T extends Record<string, infer R> ? R : never;
+ type CodomainOf<T extends Record<string, unknown>> = T extends Record<string, infer R> ? R : never;
```

@j-f1 is correct, you can use `unknown` instead of `any`.

**Comment by bradzacher** [comment]
I'm going to close and reject this for the following reasons:

- In most cases you should be able to use the much more safe `unknown` keyword.
- I don't believe there should be many (if any) use cases in your generics that _require_ the use of `any`.
- In cases that you absolutely 100% have to use any, an `eslint-disable-next-line` comment will do this job fine. because it is the edge case, there should only be a small number of them.
- Because it is the rare edge case, I don't believe there's a need to add an option which goes against the spirit of the rule, as it will mean developers can accidentally use `any` more than they absolutely need to.

**Comment by milesj** [comment]
@bradzacher While I usually agree, `unknown` does not work if the generic has a constraint. For example:

<img width="678" alt="Screen Shot 2019-07-28 at 23 00 08" src="https://user-images.githubusercontent.com/143744/62025689-c2e7ff80-b18d-11e9-9a5d-b2587cb812f1.png">

**Comment by anilanar** [comment]
@milesjs In that case, you replace unknown with `Context`.

`any` is needed when defining a generic constraint with a type that has a contravariant parameter. Once such example is `Function` and its args type which is contravariant against `Function`. There are many such cases like that because it’s a generic property of type systems. e.g. key or record/dictionary/map types, any interface whose generic parameter is used in contravariant position etc.

**Comment by anilanar** [comment]
Funny enough, tsc already knows and keeps track of what is covariant and what is contravariant, so it may be better if this would be done in typescript.

**Comment by milesj** [comment]
@anilanar `Context` cant be used here either, because it causes the `'Context' is assignable to the constraint of type 'Context', but 'Context' could be instantiated with a different subtype of constraint 'Context'.` error.

**Comment by bradzacher** [comment]
could you share more about your setup? It's hard to provide advice from a screenshot...
Maybe a better solution would be to infer the context as a generic?
```ts
export default function milesjFunc<T extends Context = Context>(
  titleOrWorkUnit: unknown,
  action?: Action<T, Input, Output>,
  scope?: unknown,
)
```

**Comment by milesj** [comment]
Probably, but that's adding far more generics all over the place for this one use case. https://github.com/milesj/boost/blob/master/packages/pipeline/src/createWorkUnit.ts#L15

I'm usually anti-any, but in cases like this it would be nice. But I'm also not a fan adding eslint line disables anywhere.

**Comment by bradzacher** [comment]
Investigating this more - you should be able to achieve this as follows:
```ts
function createWorkUnit<Input, Output = Input, TCtx extends Context = Context>(
  titleOrWorkUnit: string | WorkUnit<any, Input, Output>,
  action?: Action<TCtx, Input, Output>,
  scope?: unknown,
): WorkUnit<any, Input, Output>;
```

In usage you will never have to instantiate the generic, as its type will be inferred for you automatically based on usage:
```ts
createWorkUnit('foo'); // TCtx === Context

class FooContext extends Context {}
createWorkUnit( // TCtx === FooContext
  'foo',
  (context: FooContext, value: string, runner: any) => value,
);
```

[repl](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMIBsD2IJwEbooDeAUMsgljgPwAUAlAFzICuIA1iJgO4gDcJAL4kAJhEpwoKBNgDOYZMFkB5PACtxYZgB4AKgD5aAD2ZwQAT3rIAvPuR5MmQmYFhzABxQAlNiHyE9AEYAGmRdACY7a2Qzc1cPFAAJYGhJBAALYAQ4dBsYiwFRcXRJaRLZWWQAdUwodgBVEGAwINCItoBmOyJhEgkKtGxIIwVScgB6ACpJsnJJtCk4SBjkHG5FEHkzJGRMGGQwdOkWKClwCiGIEfyRZFl0nKx1ymwUR+R3KExPKDAU2QAdLNkJNxrMXjhaADoZIAOayUwWADaAF0mAdMpUxuRkFIwCcQBilAJyMJeld3LUFG5PMgAIIIP7YbSoMBGZBXSAgESVVCXEahACSIHcLDAoWUYtFCmiwulUWQtHB-K0aDZwVmADcciwIMw5WKNeQoL5oMwfCA-AQINoDeLkJKwPLkAAyZDJVJQDJZHIaqy2B1SsXIAA+yAACl8ALZKG2O+WFMQSKTIGBsRnAbAURaQGp1RrNW0iw2Bp3B2XF+26VnszkQbm8lV5PngTmGWZ-MCEZRQPMNJqq+RQUCw0PVWr9wuxIWViVBsD6I0xDPYajMBlMkB6Gsz6VzssLpeyGSeNesDhcXh+5h9gstafIO37hMkPo5iC3ge0ADkMEc3-oPpykqAAxRwW2GBQ6wbQZW2uHo31wXMJzvH8-0wb9QloGQ4NVMDMAgzlQm1dBdWYIcR1CE1LTNfJLBsOwSN1egBCAA)

---

The problem with adding this option is that it opens a door very widely - there's no granularity.
There's no easy way to detect if it's a "valid" use case, which means that all of a sudden every single generic type can have `any`s freely, without checks - which as I mentioned is very much against the spirit of the rule.

The point of the rule is to not use any, as there is almost always a better alternative. Very rarely should you actually have to reach for it. Some things will take a bit of work to do type-safely, but it is doable.

The point of eslint disables is to clearly document that you are working around the eslint rule for a reason. Which is exactly the case here. IMO this is completely valid code, and I wouldn't complain if I saw this in a PR:

```ts
function createWorkUnit<Input, Output = Input>(
  titleOrWorkUnit: string | WorkUnit<any, Input, Output>,
  // No point introducing a generic for this, as the type of the context doesn't matter to this function
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  action?: Action<any, Input, Output>,
  scope?: unknown,
): WorkUnit<any, Input, Output>;
```

**Comment by milesj** [comment]
I use constraints in many of my generics, so having to add additional generics to solve this ESLint rule is completely overkill. I'll just turn it off.

Thanks for the feedback.

**Comment by anilanar** [comment]
You cannot because its contravariant.

**Comment by mmun** [comment]
@bradzacher Thanks for taking the time to respond with that example. You're right. I had a fundamental misunderstanding of `unknown`.

If anyone else stumbles on this issue, check out [When to use never and unknown in TypeScript](https://blog.logrocket.com/when-to-use-never-and-unknown-in-typescript-5e4d6c5799ad/).

TL;DR:

- `unknown` is the set of all values (AKA the top type AKA the supertype of all other types)
- `never` is the empty set (AKA the bottom type AKA the subtype of all other types)
- `any` is different. It is _also_ a supertype to all other types but it has the additional (unsafe) property that the compiler will happily pretend that it is a narrower type when convenient.

**Comment by milesj** [comment]
The way I explain `unknown` to others, is that `unknown` is a type-safe alternative to `any`, and can replace `any` in most scenarios.

**Comment by bradzacher** [comment]
@milesj - an interesting thing to note that I hadn't thought about (thanks @mmun) - you can actually use `never` in place of `any` when you want to give a type to a generic which has a constraint (i.e. when you can't use `unknown`).

```ts
function createWorkUnit<Input, Output = Input>(
  titleOrWorkUnit: string | WorkUnit<never, Input, Output>,
  action?: Action<never, Input, Output>,
  scope?: unknown,
): WorkUnit<never, Input, Output>;
```
[repl](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMIBsD2IJwEbooDeAUMsgljgPwAUAlAFzICuIA1iJgO4gDcJAL4kAJhEpwoKBNgDOYZMFkB5PACtxYZgB4AKgD5aAD2ZwQAT3rIAvPuR5MmQmYFhzABxQAlNiHyE9AEYAGmRdACY7a2Qzc1cPFAAJYGhJBAALYAQ4dBsYiwFRcXRJFHx5KEQFCVlZNGwwSoQwPWQII0gQETrMdU08okE7UnJyKTgRbHRzZEx3MGA5Zi8IAEcWYCkRPX0BUYpsGGAAcxYpWjmFuWpmAAVJBZydpmQV9c2IbYMBYTEJKQoJVqyAA6pgoOwAKogYAtZTzRYgOrtTrdWZ9ZqhACSIHcLAU0TYnB4IFCynxeIJyBxlLsKIgXTqqAaTThCLkw2EJBqTIaKOQI2QAHoAFQisjkEVocaQGLIHDcRRIsBmJCzGDIMDpaRnKTgA7gflmETIWTpHJYRWUbBldC5dxQObQBYQWQAOglyBFQol1pwtDdgckx1kpgsAG0ALovLVKAWe5BSMBnECazKyPbIYRc9rucEKNyeZAAQWaiO0qDARjaHQZaOZho62Nx+LJFPxeRp+KiyFovr5HWYlaMwQlADcciwIMwu2BR2NfNBlr5-BBtLO22BacgAGTIZKpKAZLI5UdWWzIclbjsAH2Qt0dAFslGur7TCn8SgCYGwy9gKDKEBghC0KwuuLZzpe7ZUrOhgSgsYCEMoUDAVCMJaKajSgMcyB3qhoEtDgY7QM2lKbrS84xH+IA3CW1HaERJHUhB5HdpRsgyJ4tFElwvBnsw+HoQxEDEVApGtlB15gLsJDcoBgmwrQADkMCOEp9DckCdQAGKOA2kAdDWqK8o2CiDNycgKNW0QIPJ4JoYpKlqaEtAyKZzC6Zg+koqEE7oFOzAVNhoRQIuUBhpYNh2H5U70IURhum5RynOcgxxSQQA)

**Comment by milesj** [comment]
@bradzacher Oh wow, never would of though about that either. I'll give that shot. For a few of those scenarios, I've been using `{}` which has done the job also.
